endgame_models/get_differences.py

```diff
@@ -0,0 +1,99 @@
+from dataclasses import dataclass
+from typing import Optional, TypeVar
+
+from typing_extensions import Type
+
+from .models import BaseInitialParams
+
+Flat = str | float | int | bool
+
+JSONType = Flat | list["JSONType"] | dict[str, "JSONType"] | tuple["JSONType", ...]
+DictType = dict[str, JSONType]
+ListType = list[JSONType]
+TupleType = tuple[JSONType, ...]
+
+
+@dataclass
+class ReadOnlyDiff:
+    original: JSONType
+    new: JSONType
+
+
+OutputReadOnly = dict[str, "ReadOnlyDiff | OutputReadOnly"]
+
+
+def _output_read_only_diff(
+    initial_dict: DictType, update_dict: DictType, model: Type[BaseInitialParams]
+) -> Optional[OutputReadOnly]:
+    changed_read_only: OutputReadOnly = {}
+    for k, new_v in update_dict.items():
+        if (old_v := initial_dict.get(k)) is not None:
+            assert k in model.__fields__
+            # Update and initial both have parameter
+            field = model.__fields__[k]
+            if not field.field_info.allow_mutation:
+                if issubclass(field.outer_type_, BaseInitialParams):
```

Comment:
What's `outer_type_` is it some hidden feature? Are you ok to use it?

Reply:
This whole thing relies on hidden features haha - outer_type_ is preferable to type_ - I remembered it is essentially the type you see in the class, whereas type_ is... inconsistent